### PR TITLE
fix: dual-stack address normalisation and NAT traversal deadlock

### PR DESCRIPTION
**Pull request by jacderida**
## Summary
- Normalise dual-stack IPv4-mapped addresses across all NAT traversal lookup paths
- Fix DashMap + parking_lot deadlock in `poll()` that caused tokio runtime livelock
- Fix connection lifecycle issues preventing local node from responding to NAT-traversed peers

### Deadlock fix
`poll()` used `active_sessions.iter_mut()` which holds DashMap WRITE guards on ALL shards while calling `discovery_manager.lock()`. When the discovery poll task contended on the same Mutex, any third task touching `active_sessions` (e.g., `initiate_nat_traversal`) deadlocked the runtime. Replaced with snapshot-and-update pattern using per-key `get_mut()`.

### Connection lifecycle fixes
- Reader task registers its connection in the DashMap so the send path can find it
- `add_connection()` and `attempt_hole_punch_connection()` check for existing live connections before inserting
- `remove_connection()` checks `close_reason` before removing, preserving live connections
- Removed fire-and-forget QUIC probe that clobbered incoming connections

### Dual-stack address normalisation
- `dual_stack_alternate()` helper tries both plain and mapped address forms
- Applied to PUNCH_ME_NOW relay lookup, address validation, wire ID encoding, and `P2pEndpoint::send()`

## Test plan
- [x] 60-minute sustained upload test on 6-node testnet (5 cloud + 1 local behind NAT)
- [x] 43 consecutive uploads, 0 failures, 357 chunks stored on local node
- [x] No deadlocks, no process hangs, no runtime starvation

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR addresses three intertwined reliability issues in the NAT traversal stack: a DashMap + `parking_lot` deadlock in `poll()`, connection lifecycle races that prevented hole-punched peers from sending/receiving data, and inconsistent dual-stack address normalisation that caused lookup misses on dual-stack Linux sockets.\n\nThe deadlock fix (snapshot-and-update replacing `iter_mut()`) is sound. The connection lifecycle changes — guarding `add_connection` and `remove_connection` with `close_reason()` checks, removing the fire-and-forget QUIC probe, and unifying the accept path through `spawn_accept_loop` + `accept_connection_direct` — address the root causes described in the PR. The dual-stack normalisation helpers (`normalize_socket_addr`, `dual_stack_alternate`, `wire_id_from_addr`) are clean and well-placed in `shared.rs`.\n\nKey concerns:\n\n- **Hot-path `info!` logging** (`p2p_endpoint.rs`): Multiple `trace!`/`debug!` calls were promoted to `info!`, including one that fires on every received message. At production throughput this would overwhelm log aggregation and add non-trivial overhead.\n- **Dead diagnostic variable** (`p2p_endpoint.rs`): `peers_has_alt` is hardcoded to `false` via a closure that always returns `false`, yet is still computed and referenced in a nearby comment as if it has meaning.\n- **Potential duplicate reader tasks** (`nat_traversal_api.rs`): The inner handshake task inside `spawn_accept_loop` ignores the return value of `emitted2.insert()` and unconditionally sends to `handshake_tx`. In simultaneous-open scenarios (normal for NAT traversal), both the `incoming_notify` path and the handshake path can deliver the same address to `handshake_tx`, causing `P2pEndpoint::accept()` to spawn two reader tasks.\n- **`remote_address()` semantic change** (`high_level/connection.rs`): Switched from a live Quinn lock query to a cached initial address to avoid `parking_lot` contention. The doc comment is updated, but callers that expect the post-migration address will now silently get stale data without any compile-time indication.

<h3>Confidence Score: 4/5</h3>

Safe to merge after addressing the hot-path info! logging; the logic issues are edge-cases in simultaneous-open that don't affect the primary NAT traversal path validated by the 6-node testnet.

The core fixes (deadlock, connection lifecycle, dual-stack normalisation) are well-targeted and backed by a 60-minute sustained test with zero failures. The P1 logging concern is important for production but doesn't affect correctness. The duplicate-reader-task issue is a real edge case bounded to simultaneous-open scenarios and results in wasted resources rather than data loss. No data-loss or security issues were found.

src/p2p_endpoint.rs (hot-path info! logging, peers_has_alt dead code) and src/nat_traversal_api.rs (emitted2.insert return value ignored in spawn_accept_loop)

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/p2p_endpoint.rs | Core endpoint with deadlock fix, dual-stack send(), and new fallback connection logic; contains per-message info! logging and a dead peers_has_alt variable |
| src/nat_traversal_api.rs | Replaces old accept_connections spawn with unified spawn_accept_loop + accept_connection_direct; inner handshake task ignores emitted.insert() return value, risking duplicate handshake_tx sends |
| src/high_level/connection.rs | remote_address() now returns initial_remote_addr instead of live Quinn address; observed_address() switched to try_lock to avoid blocking; both changes are intentional but change contract |
| src/shared.rs | Adds dual_stack_alternate() and deterministic wire_id_from_addr() helpers; clean and well-documented implementation |
| src/connection/nat_traversal.rs | Adds stale-coordination reset, dual-stack addr normalisation, and auto-start of coordination round for relayed PUNCH_ME_NOW; should_reset logic is correct but unnecessarily indirect |
| src/frame/nat_traversal_unified.rs | target_peer_id now encoded as an extension byte after standard RFC fields; backwards-compatible with correct round-trip tests updated |
| src/high_level/endpoint.rs | Hole-punch address processing moved from fire-and-forget to channel-based tracking; default_client_config and hole_punch_tx added to State |
| src/high_level/mutex.rs | Adds try_lock() to both tracking and non-tracking Mutex wrappers; straightforward implementation |
| src/connection/mod.rs | PUNCH_ME_NOW handling now guards coordinator path with target_peer_id.is_some(); removes fire-and-forget QUIC probe, replaced with InitiateHolePunch event |
| src/link_transport_impl.rs | dial_addr now uses connect_with_fallback with happy-eyeballs; connected address is resolved post-connect for relay/hole-punch scenarios |

</details>

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant NATNode as NAT Node (local)
    participant Coord as Coordinator (bootstrap)
    participant Cloud as Cloud VM (target)

    Note over NATNode: try_hole_punch(target=Cloud, coordinator=Coord)
    NATNode->>Coord: PUNCH_ME_NOW (target_peer_id=Cloud, round=R)
    Note over Coord: connection/mod.rs: target_peer_id.is_some() guard
    Coord->>Cloud: RelayPunchMeNow(Cloud, frame, sender_addr)

    Note over Cloud: nat_traversal_api.rs: handle_endpoint_event
    Cloud->>Cloud: attempt_hole_punch_connection(NATNode.addr)
    Cloud->>Cloud: spawn connection task → DashMap.insert(NATNode, conn)
    Cloud->>Cloud: accepted_addrs_tx.send(NATNode.addr)
    Cloud->>Cloud: incoming_notify.notify_waiters()

    Note over Cloud: spawn_accept_loop: incoming_notify fires
    Cloud->>Cloud: emitted.insert(NATNode) → true → handshake_tx.send()
    Cloud->>Cloud: P2pEndpoint::accept() → spawn_reader_task(NATNode)

    Cloud->>NATNode: QUIC Initial (hole-punch)
    Note over NATNode: spawn_accept_loop: endpoint.accept()
    NATNode->>NATNode: Handshake completes → DashMap.insert(Cloud, conn)
    NATNode->>NATNode: P2pEndpoint::accept() → spawn_reader_task(Cloud)
    NATNode->>NATNode: connected_peers.insert(Cloud, peer_conn)

    Note over NATNode: try_hole_punch poll: inner.is_connected(Cloud) → true
    NATNode-->>NATNode: return Ok(peer_conn)

    NATNode->>Cloud: open_uni → send data
    Cloud->>NATNode: open_uni → send data (via reader task)
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: src/p2p_endpoint.rs
Line: 2596

Comment:
**Per-message `info!` log floods production logs**

This line was previously `tracing::trace!(...)` and was changed to `info!`. On a sustained upload run (the PR's own test reports 357 chunks), this emits an `INFO`-level log entry for every received message, which would flood any log aggregation system and add measurable overhead on the receiver hot path.

Or simply revert the level:

```suggestion
                tracing::trace!("Reader task: {} bytes from {}", data_len, addr);
```

The broader pattern of promoting many hot-path `trace!`/`debug!` calls to `info!` throughout this PR (reader task lifecycle, `try_hole_punch` poll iterations, `inner_is_connected` every call) is worth auditing before merge. Diagnostic verbosity is valuable for debugging, but `info` is typically enabled in production.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: src/p2p_endpoint.rs
Line: 2163-2167

Comment:
**`peers_has_alt` is always `false` — dead diagnostic variable**

The closure passed to `map_or` unconditionally returns `false`, so `peers_has_alt` is always `false` regardless of whether the alternate address is in `connected_peers`. The variable is also never referenced in the subsequent `if !peers_has && !inner_has && !inner_has_alt` condition, making it entirely dead code. The comment "Can't read while holding read" is misleading because the `connected_peers.read()` guard was already released by the time this binding is evaluated (it was a temporary).

Since this whole block is diagnostic-only and the `inner_has_alt` check already covers the important case, remove `peers_has_alt` entirely to avoid confusion.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: src/high_level/connection.rs
Line: 669-670

Comment:
**`remote_address()` now returns a stale value after connection migration**

Changing this from `self.0.state.lock("remote_address").inner.remote_address()` to `self.0.initial_remote_addr` avoids a lock acquisition — but silently breaks callers that expect the *current* remote address after a QUIC migration event.

The updated doc comment documents the limitation, which helps. However, this is a semantic contract change on a public method. It would be worth confirming that no path in the NAT-traversal or connection-lifecycle code depends on `remote_address()` returning the post-migration address. If migration is unconditionally disabled for P2P nodes in this codebase, a compile-time assertion or a `debug_assert` would make that explicit and prevent future regressions.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: src/nat_traversal_api.rs
Line: 3587-3640

Comment:
**`spawn_accept_loop` inner handshake task unconditionally sends to `handshake_tx`, risking duplicate reader tasks**

The `incoming_notify` branch gates sends on `emitted.insert(addr)` (returns `true` only on first insert). But the inner handshake task calls `emitted2.insert(remote_address)` *without checking the return value*, then always calls `tx2.send(Ok((remote_address, connection))).await`:

```rust
emitted2.insert(remote_address);  // return value ignored
// ...
let _ = tx2.send(Ok((remote_address, connection))).await;
```

In a simultaneous-open scenario (NAT traversal — both peers connect to each other at the same time):
1. `add_connection(X)` is called (outgoing side) → `incoming_notify` fires
2. `incoming_notify` handler: `emitted.insert(X)` returns `true` → sends `(X, outgoing_conn)` to `handshake_tx`
3. Quinn also accepts the incoming connection from X in the next `endpoint.accept()` cycle
4. Handshake task for that accept completes → `emitted2.insert(X)` returns `false` (already inserted in step 2), but `tx2.send(Ok((X, incoming_conn)))` is still called

Result: `handshake_tx` has two entries for `X`. `P2pEndpoint::accept()` processes both, spawning two reader tasks for the same connection address.

The fix is straightforward — only send to `handshake_tx` if this is the first time the address was emitted:

```rust
if emitted2.insert(remote_address) {
    let _ = tx2.send(Ok((remote_address, connection))).await;
}
```

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix: NAT traversal deadlock, connection ..."](https://github.com/saorsa-labs/saorsa-transport/commit/1b543ff6d578884dc547bc7e7de5c3dfff6fcbbf) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=26498553)</sub>

> Greptile also left **4 inline comments** on this PR.

### CODE REVIEW

File: src/high_level/connection.rs [hunk 1]
```diff
@@ -646,20 +649,25 @@ impl Connection {
...
     ///
-    /// If `ServerConfig::migration` is `true`, clients may change addresses at will, e.g. when
-    /// switching to a cellular internet connection.
+    /// Note: this returns the address captured when the connection was first
+    /// established. If `ServerConfig::migration` is `true`, the peer may have
+    /// migrated to a different address since then. This value does not update
+    /// after migration.
     pub fn remote_address(&self) -> SocketAddr {
-        self.0.state.lock("remote_address").inner.remote_address()
+        self.0.initial_remote_addr
```
[the remark this answers is not in the document]
Comment (jacderida) [reply src/high_level/connection.rs]: This change is from a prior commit (e7776f34) that predates the NAT traversal work in this PR. Connection migration is not used in our P2P architecture — all connections are established to fixed endpoints. Agree that a debug_assert would be a good safeguard but will address that separately.

**Comment by mickvandijke** [comment]
Replaced by https://github.com/saorsa-labs/saorsa-transport/pull/27
